File: server/projects/profile/src/main/java/com/raxim/myscoutee/profile/service/ProfileService.java
package com.raxim.myscoutee.profile.service;

import java.io.IOException;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Stream;

import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import com.raxim.myscoutee.common.util.CommonUtil;
import com.raxim.myscoutee.common.util.SettingUtil;
import com.raxim.myscoutee.profile.data.document.mongo.Event;
import com.raxim.myscoutee.profile.data.document.mongo.Profile;
import com.raxim.myscoutee.profile.data.dto.rest.GroupDTO;
import com.raxim.myscoutee.profile.data.dto.rest.PageParam;
import com.raxim.myscoutee.profile.data.dto.rest.ProfileDTO;
import com.raxim.myscoutee.profile.data.dto.rest.SettingDTO;
import com.raxim.myscoutee.profile.exception.MessageException;
import com.raxim.myscoutee.profile.repository.mongo.EventRepository;
import com.raxim.myscoutee.profile.repository.mongo.ProfileEventHandler;
import com.raxim.myscoutee.profile.repository.mongo.ProfileRepository;
import com.raxim.myscoutee.profile.repository.mongo.UserRepository;
import com.raxim.myscoutee.profile.util.AppConstants;
import com.raxim.myscoutee.profile.util.ProfileUtil;

//val reqGroupId = UUID.fromString(groupId)
//if (profile.group == reqGroupId) {
/*val role = this.roleRepository.findRoleByProfileAndGroup(adminEmail, groupId = reqGroupId)
    .firstOrNull { role -> role.group.id == reqGroupId && role.role == FirebaseService.ROLE_ADMIN && role.status == "A" }*/
//}

@Service
public class ProfileService {

    private final ProfileRepository profileRepository;
    private final EventRepository eventRepository;
    private final ProfileEventHandler profileEventHandler;
    private final UserRepository userRepository;
    private final SettingsService settingsService;

    public ProfileService(
            ProfileRepository profileRepository,
            EventRepository eventRepository,
            ProfileEventHandler profileEventHandler,
            UserRepository userRepository,
            SettingsService settingsService) {
        this.profileRepository = profileRepository;
        this.eventRepository = eventRepository;
        this.profileEventHandler = profileEventHandler;
        this.userRepository = userRepository;
        this.settingsService = settingsService;
    }

    public List<ProfileDTO> getProfiles(PageParam pageParam, Profile profile, String pKey, double direction,
            int score)
            throws MessageException {

        if (profile.getPosition() == null) {
            throw new MessageException(AppConstants.ERR_NO_PROFILE);
        }

        if ("F".equals(profile.getStatus())) {
            throw new MessageException(AppConstants.ERR_FRIENDS_ONLY);
        }

        if ("I".equals(profile.getStatus())) {
            throw new MessageException(AppConstants.ERR_INVISIBLE_PROFILE);
        }

        // gender,met,direction as filter criteria
        // based on pKey it can be different, maybe create ParamParserForDouble etc.
        Optional<SettingDTO> optSetting = this.settingsService.getSetting(profile.getId(), pKey);

        Optional<String> optMet = SettingUtil.getValue(optSetting, "met");
        boolean met = optMet.isPresent() ? Boolean.parseBoolean(optMet.get()) : false;

        Optional<String> optGender = SettingUtil.getValue(optSetting, "gender");

        Optional<String> optDirection = SettingUtil.getValue(optSetting, "direction");
        direction = optDirection.isPresent() ? Double.parseDouble(optDirection.get()) : direction;

        String gender = null;
        UUID selectUuid = pageParam.getSelectedId() != null ? UUID.fromString(pageParam.getSelectedId()) : null;
        if (selectUuid != null) {
            Optional<Profile> selected = profileRepository.findById(selectUuid);
            if (!selected.isPresent()) {
                throw new MessageException(AppConstants.ERR_NO_PROFILE);
            }

            Profile sProfile = selected.get();
            gender = sProfile.getGender().equals("m") ? "w" : "m";
        } else {
            gender = optGender.isPresent() ? optGender.get() : profile.getGender().equals("m") ? "w" : "m";
        }

        // $met variable is a filter for findProfile on the list screen,
        // can be true or false (events are not even generated, but both liked)
        // isDouble = true, direction is ignored
        List<ProfileDTO> profileDTOs = profileRepository.findProfile(
                pageParam,
                CommonUtil.point(profile.getPosition()),
                gender,
                profile.getGroup(),
                direction, score, met, selectUuid);
        return profileDTOs;
    }

    public List<GroupDTO> getGroupsByProfile(UUID profileId, PageParam pageParam) {
        return this.profileRepository.findGroupsByProfile(profileId, pageParam);
    }

    public Optional<ProfileDTO> saveProfile(String profileId,
            Profile pProfile) {
        UUID pProfileUuid = UUID.fromString(profileId);
        Optional<Profile> dbProfile = profileRepository.findById(pProfileUuid);

        if (dbProfile.isPresent()) {
            Profile profile = dbProfile.get();
            // ??? - default is U, and admin of the group can change it
            profile.setRole(pProfile.getRole());
            profile.setStatus(pProfile.getStatus());

            List<Event> events = this.eventRepository.findAll();

            List<Event> sEvents = null;
            // friends only
            if ("F".equals(pProfile.getStatus())) {
                sEvents = changeStatus(profile, events.stream()
                        .filter(event -> "P".equals(event.getStatus()) && "A".equals(event.getAccess())), "LF");
            } else if ("I".equals(pProfile.getStatus())) {
                sEvents = changeStatus(profile, events.stream()
                        .filter(event -> "P".equals(event.getStatus()) && !"P".equals(event.getAccess())), "LI");
            } else if ("S".equals(pProfile.getStatus())) {
                sEvents = changeStatus(profile, events.stream()
                        .filter(event -> "P".equals(event.getStatus())), "LS");
            } else if ("L".equals(pProfile.getStatus())) {
                sEvents = changeStatus(profile, events.stream(), "LG");
            }
            this.eventRepository.saveAll(sEvents);

            Profile profileSaved = profileRepository.save(profile);
            return Optional.of(new ProfileDTO(profileSaved));
        }
        return Optional.empty();
    }

    private List<Event> changeStatus(Profile profile, Stream<Event> sEvents, String toStatus) {
        return sEvents.map(event -> {
            event.getMembers().stream().filter(member -> profile.getId().equals(member.getProfile().getId()))
                    .map(member -> {
                        member.setStatus(toStatus); // friends only leave
                        return member;
                    });
            return event;
        }).toList();
    }

    public ProfileDTO saveProfile(UUID userId, UUID profileId, UUID group,
            Profile profile, MultipartFile voice) throws IOException {

        String fileName = UUID.randomUUID().toString();

        Optional<Profile> profileResult = profileId != null ? profileRepository.findById(profileId) : Optional.empty();

        Profile clonedProfile = profile.clone();
        if (profileResult.isPresent()) {
            Profile profileEntity = profileResult.get();

            clonedProfile.setId(profileEntity.getId());
            clonedProfile.setCars(profileEntity.getCars());
            clonedProfile.setSchools(profileEntity.getSchools());
            if (profileEntity.getVoice() == null) {
                clonedProfile.setVoice(fileName);
            }
            clonedProfile.setGroup(group);
        } else {
            clonedProfile.setId(UUID.randomUUID());
            clonedProfile.setVoice(fileName);
            clonedProfile.setGroup(group);
        }
        clonedProfile.setScore(ProfileUtil.score(clonedProfile));

        ProfileUtil.saveVoice(voice, clonedProfile.getVoice());

        Profile profileSaved = this.profileRepository.save(clonedProfile);
        this.userRepository.addProfile(userId, profileSaved);

        ProfileDTO profileDTO = new ProfileDTO();
        profileDTO.setProfile(profileSaved);
        return profileDTO;
    }
}

File: server/projects/profile/src/main/java/com/raxim/myscoutee/profile/service/ProfileService.kt
package com.raxim.myscoutee.profile.service

import com.raxim.myscoutee.common.FileUtil
import com.raxim.myscoutee.common.save
import com.raxim.myscoutee.profile.controller.score
import com.raxim.myscoutee.profile.data.document.mongo.Profile
import com.raxim.myscoutee.profile.data.dto.rest.Car
import com.raxim.myscoutee.profile.data.dto.rest.ProfileStatus
import com.raxim.myscoutee.profile.data.dto.rest.School
import com.raxim.myscoutee.profile.repository.mongo.*
import org.springframework.stereotype.Service
import org.springframework.web.multipart.MultipartFile
import java.io.ByteArrayInputStream
import java.io.InputStream
import java.util.*

@Service
class ProfileService(
    private val profileRepository: ProfileRepository,
    private val profileEventHandler: ProfileEventHandler,
    private val carRepository: CarRepository,
    private val carEventHandler: CarEventHandler,
    private val schoolRepository: SchoolRepository,
    private val userRepository: UserRepository,
    private val roleRepository: RoleRepository
) {

    //val reqGroupId = UUID.fromString(groupId)
    //if (profile.group == reqGroupId) {
    /*val role = this.roleRepository.findRoleByProfileAndGroup(adminEmail, groupId = reqGroupId)
        .firstOrNull { role -> role.group.id == reqGroupId && role.role == "ROLE_ADMIN" && role.status == "A" }*/
    //}

    fun saveProfileStatus(
        profileId: String,
        profileStatus: ProfileStatus
    ): com.raxim.myscoutee.profile.data.dto.rest.Profile? {
        return profileRepository.findById(UUID.fromString(profileId)).map { profile ->
            val profileToSave = profile.copy(status = profileStatus.status)
            val profileSaved = profileRepository.save(profileToSave)

            val role = this.roleRepository.findRoleByProfile(profileSaved.id!!)

            val firstRole = role.first().copy(role = profileStatus.role!!)
            this.roleRepository.save(firstRole)

            com.raxim.myscoutee.profile.data.dto.rest.Profile(profile = profileSaved)
        }.get()
    }

    fun saveProfile(
        userId: UUID,
        profileId: UUID?,
        group: UUID,
        profile: Profile,
        voice: MultipartFile? = null
    ): com.raxim.myscoutee.profile.data.dto.rest.Profile? {
        val fileName: String = UUID.randomUUID().toString()

        val (separator, permDir) = FileUtil.uuidToPath(fileName, true)

        if (voice != null && !voice.isEmpty) {
            val voiceStream: InputStream = ByteArrayInputStream(voice.bytes)
            voiceStream.save("$permDir${separator}_${fileName}_voice")
        }

        val profileToSave = if (profileId != null) {
            val profileResult = this.profileRepository.findById(profileId)
            val profileEntity = profileResult.get()
            profile.copy(
                id = profileEntity.id,
                cars = profileEntity.cars,
                schools = profileEntity.schools,
                voice = fileName,
                group = group
            )
        } else {
            profile.copy(
                id = UUID.randomUUID(),
                voice = fileName,
                group = group
            )
        }

        val profileSaved = this.profileRepository.save(profileToSave.copy(score = profileToSave.score()))

        this.userRepository.findById(userId).map { user ->
            user.profile = profileSaved

            user.profiles!!.add(profileSaved)
            userRepository.save(user)
        }

        this.profileEventHandler.handleAfterCreate(profileSaved)
        return com.raxim.myscoutee.profile.data.dto.rest.Profile(profileSaved)
    }

    fun getCars(profileId: UUID, step: Int?, tOffset: Array<Any>): List<Car> {
        return this.profileRepository.findCarsByProfile(
            profileId, 20, step ?: 5,
            tOffset
        )
    }

    fun getCarByProfile(profileId: UUID, carId: UUID): Optional<com.raxim.myscoutee.profile.data.document.mongo.Car> {
        return this.profileRepository.findCarByProfile(
            profileId, carId
        )
    }

    fun getSchools(profileId: UUID, step: Int?, tOffset: Array<Any>): List<School> {
        return this.profileRepository.findSchoolsByProfile(
            profileId, 20, step ?: 5,
            tOffset
        )
    }

    fun getSchoolByProfile(
        profileId: UUID,
        schoolId: UUID
    ): Optional<com.raxim.myscoutee.profile.data.document.mongo.School> {
        return this.profileRepository.findSchoolByProfile(
            profileId, schoolId
        )
    }

    fun addCar(profileId: UUID, carId: UUID?, newCar: com.raxim.myscoutee.profile.data.document.mongo.Car): Car {
        return this.profileRepository.findById(profileId).map { profile ->

            val car = if (carId != null) {
                val dbCar = profile.cars!!.first { car -> car.id == carId }
                newCar.copy(id = dbCar.id)
            } else {
                newCar
            }

            val savedCar = carRepository.save(car)
            profile.cars!!.add(savedCar)
            profileRepository.save(profile)
            carEventHandler.handleAfterCreate(car)
            Car(savedCar)
        }.get()
    }

    fun saveSchools(
        profileId: UUID,
        schools: List<com.raxim.myscoutee.profile.data.document.mongo.School>
    ): List<School> {
        return this.profileRepository.findById(profileId).map { profile ->

            for (school in schools) {
                val savedSchool = this.schoolRepository.save(school)
                val hasSchool = profile.schools!!.any { school -> school.id == savedSchool.id }
                if (!hasSchool) {
                    profile.schools.add(savedSchool)
                }
            }
            this.profileRepository.save(profile)

            profile.schools!!.map { school ->
                School(school)
            }
        }.get()
    }
}